src/coprocessor/dag/expr.rs

```diff
@@ -0,0 +1,238 @@
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Io(err: io::Error) {
+            from()
+            description("io error")
+            display("I/O error: {}", err)
+            cause(err)
+        }
+        Type { has: &'static str, expected: &'static str } {
+            description("type error")
+            display("type error: cannot get {:?} result from {:?} expression", expected, has)
+        }
+        Codec(err: util::codec::Error) {
+            from()
+            description("codec error")
+            display("codec error: {}", err)
+            cause(err)
+        }
+        ColumnOffset(offset: usize) {
+            description("column offset not found")
+            display("illegal column offset: {}", offset)
+        }
+        Other(desc: &'static str) {
+            description(desc)
+            display("error {}", desc)
+        }
+    }
+}
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+/// Flags are used by `SelectRequest.flags` to handle execution mode, like how to handle
+/// truncate error.
+/// `FLAG_IGNORE_TRUNCATE` indicates if truncate error should be ignored.
+/// Read-only statements should ignore truncate error, write statements should not ignore
+/// truncate error.
+pub const FLAG_IGNORE_TRUNCATE: u64 = 1;
+/// `FLAG_TRUNCATE_AS_WARNING` indicates if truncate error should be returned as warning.
+/// This flag only matters if `FLAG_IGNORE_TRUNCATE` is not set, in strict sql mode, truncate error
+/// should be returned as error, in non-strict sql mode, truncate error should be saved as warning.
+pub const FLAG_TRUNCATE_AS_WARNING: u64 = 1 << 1;
+
+/// StatementContext contains variables for a statement
+/// It should be reset before executing a statement
+pub struct StatementContext {
+    ignore_truncate: bool,
+    truncate_as_warning: bool,
+    timezone: FixedOffset,
+}
+
+/// Expression represents all scalar expression in SQL.
+#[derive(Debug, Clone, PartialEq)]
+pub enum Expression {
+    Constant(Datum),
+    ColumnRef(usize),
+    ScalarFn(FnCall),
+}
+
+/// A single scalar function call
+#[derive(Debug, Clone, PartialEq)]
+pub struct FnCall {
+    sig: ScalarFuncSig,
+    children: Vec<Expression>,
+}
+
+impl Datum {
+    fn to_i64(&self) -> Option<i64> {
+        match *self {
+            Datum::Null => None,
+            _ => Some(self.i64()),
+        }
+    }
+
+    fn to_f64(&self) -> Option<f64> {
+        match *self {
+            Datum::Null => None,
+            _ => Some(self.f64()),
+        }
+    }
+}
+
+impl Expression {
+    fn eval_int(&self, row: &[Datum], ctx: &StatementContext) -> Result<Option<i64>> {
+        match *self {
+            Expression::Constant(ref datum) => Ok(datum.to_i64()),
+            Expression::ColumnRef(offset) => {
+                row.get(offset).map(Datum::to_i64).ok_or_else(|| Error::ColumnOffset(offset))
+            }
+            Expression::ScalarFn(ref fun) => {
+                match fun.sig {
```

Comment:
Shoud we move this to a public place? `eval_real` and other `eval_*` will use this, too.